### PR TITLE
feat(media): write per-upload-event records for moderation

### PR DESCRIPTION
**Pull request by baxen**
## Summary

Adds env-gated per-event records at `_uploads/{community_id}/{sha256}/{event_id}.json` in the media bucket — one per accepted upload, ULID event id. 

## Record shape (version 1)

Versioned/extensible JSON with everything the moderation pipeline needs with **no further lookups**:

```json
{
  "version": 1,
  "event_id": "01J...",            // ULID, also the object key suffix
  "sha256": "...",
  "ext": "png",
  "mime_type": "image/png",
  "size": 12345,
  "uploaded_at": 1751820000,
  "community_id": "...",
  "community_host": "example.buzz.host",
  "uploader_id": "<hex pubkey>",
  "uploader_npub": "npub1...",
  "uploader_name": "display name at upload time",  // optional
  "ip": "203.0.114.7",                              // optional, opt-in
  "port": 51234                                      // optional, only with ip
}
```

## Design decisions

- **Record written LAST, and record-write failure fails the upload.** Blob + thumbnail + sidecar are durably stored before the record is PUT. The record's S3 `ObjectCreated` notification is the moderation scan trigger, so record existence must imply everything it references is readable. This makes a single record-driven consumer path sufficient — no legacy blob-event path needed.
- **Idempotent short-circuit also writes a record.** Re-uploading known bytes does no blob PUT; without this, takedown re-uploads would be invisible to moderation. Each re-upload is a distinct event record under the same `{sha256}/` prefix.
- **IP+port are opt-in and fail-empty** (NCMEC-suitable for operators with legal obligations; Buzz collects nothing by default). IP is read only from the operator-configured trusted edge header (`BUZZ_MEDIA_UPLOAD_IP_HEADER`, e.g. `cf-connecting-ip`), validated as a public address (v4+v6; private/loopback/CGNAT/documentation ranges rejected). The socket address is never used — behind a sidecar it's meaningless, and a wrong address is worse than none. Port kept only alongside a valid IP.
- **Fail-loud config validation at startup**: IP header set without records enabled, port header without IP header, or invalid header-name syntax are startup errors — an operator setting the IP header believes they're meeting a legal obligation, so silent no-op is unacceptable.
- **Records are never client-readable**: the media serve path rejects `_uploads/` and `_meta/` key shapes (tested).

## Configuration

| Env var | Helm value | Default |
|---|---|---|
| `BUZZ_MEDIA_UPLOAD_RECORDS` | `relay.uploadRecords` | off |
| `BUZZ_MEDIA_UPLOAD_IP_HEADER` | `relay.uploadIpHeader` | unset (no IP collection) |
| `BUZZ_MEDIA_UPLOAD_PORT_HEADER` | `relay.uploadPortHeader` | unset |

## Testing

- Unit tests in `upload_record.rs` (key shape, serialization round-trip, public-IP validation incl. v6 edge cases, port parsing) and `config.rs` (validation coherence matrix)
- Serve-path guard test in `api/media.rs`
- `cargo test -p buzz-media` (53 passed) and `-p buzz-relay` (478 passed); `cargo check --locked --workspace` clean
- Cargo.lock delta is minimal: only the `ulid` package addition

### CODE REVIEW

File: crates/buzz-media/src/upload.rs [hunk 1]
```diff
@@ -105,15 +133,37 @@ where
...
-            body.len() as u64,
-            Some(&meta),
-            uploaded_at,
-        )),
+        Ok(meta) => {
+            // Per-event record LAST — blob, thumbnail, and sidecar are all
+            // durably stored, so the record's ObjectCreated event (the
+            // moderation scan trigger) implies everything it references is
+            // readable.
+            if let Some(attribution) = &attribution {
```
Comment (wpfleger96) [inline crates/buzz-media/src/upload.rs]: hey Brad — I think there's a real hole in the fresh-upload ordering here. The record is written last, after the sidecar is already committed inside `generate_and_store_metadata` (`put_sidecar` at the tail). But the serve path gates *only* on the sidecar — `get_blob` does `read_sidecar_mime(...).ok_or(NotFound)?` in `media.rs` and never looks at the `_uploads/` record (`head_blob` is the same). So if the record PUT fails here, the sidecar's already durable and the blob is fully servable by hash, but the record's `ObjectCreated` event — the only moderation-scan trigger — never fired. That's the exact case this feature exists to prevent: publishable media that skipped the hook. A retry re-drives the idempotent branch and can write the record, but nothing forces a client to retry, and a bad `_uploads/` prefix policy would make every upload error out while still publishing via the sidecar. Could we make the record part of the publish gate — either hold the sidecar write until the record succeeds, or have `get_blob`/`head_blob` require the record? The record already carries `ext`/`mime_type`/`size` so reordering looks viable. Same ordering applies on the video path further down (sidecar then record).

File: crates/buzz-media/src/upload.rs [hunk 2]
```diff
@@ -31,12 +32,19 @@ use crate::validation::{
 /// key, the both-exist idempotency short-circuit, blob store, orphan-blob
 /// handling, and descriptor build — is common. The streaming video path stays
 /// separate (see [`process_video_upload`]) because it never buffers in RAM.
+///
+/// `attribution` is `Some` when per-event upload records are enabled
+/// (`BUZZ_MEDIA_UPLOAD_RECORDS`): a record is then written for **every**
+/// accepted upload — including the idempotent short-circuit, which does no
+/// blob PUT and would otherwise be invisible to the moderation pipeline. The
+/// record is written last, after all other objects are durably stored.
 async fn process_buffered_upload<V, M, Fut>(
```
Comment (wpfleger96) [inline crates/buzz-media/src/upload.rs]: CI's red on this one — `clippy::too_many_arguments` (8/7) at the signature, and Rust Lint / Windows Rust run `-D warnings` so it's a hard fail, not just a warning. Rather than `#[allow(...)]`, I think the cleaner fix is bundling the six stable request-context params (`storage`, `config`, `ctx`, `auth_event`, `body`, `attribution`) into a small internal `BufferedUploadInput<'a>` and leaving `validate` + `store_metadata` as the two real strategy hooks — it makes the invariant-vs-strategy split explicit instead of papering over it. Your call on the shape, but something that clears `-D warnings` structurally.

File: crates/buzz-media/src/upload_record.rs [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+    match ip {
+        IpAddr::V4(v4) => {
+            let octets = v4.octets();
+            !(v4.is_private()
+                || v4.is_loopback()
+                || v4.is_link_local()
+                || v4.is_broadcast()
+                || v4.is_documentation()
+                || v4.is_multicast()
+                || v4.is_unspecified()
```
Comment (wpfleger96) [inline crates/buzz-media/src/upload_record.rs:217]: one thing on the public-IP check: the v4 branch rejects only `0.0.0.0` via `is_unspecified()`, so an address like `0.1.2.3` in `0.0.0.0/8` (IANA "this-network", not globally reachable) gets recorded as a public IP. The v6 branch just below is also missing a few non-globally-reachable ranges — 6to4 `2002::/16`, Teredo `2001::/32`, benchmarking `2001:2::/48`, documentation `3fff::/20`, discard `100::/64`. Given the module's own note that a false positive is "a wrong address in a federal report," the false-positive direction is the one that actually hurts here, so I'd tighten it: reject `0.0.0.0/8` on v4 and add a v6 deny table for those special ranges, with tests. Not blocking the merge but I'd like it in before this ships enabled anywhere.

File: crates/buzz-media/src/config.rs [hunk 4]
```diff
@@ -72,6 +88,100 @@ impl MediaConfig {
...
+                    .to_string(),
+            );
+        }
+        for (name, value) in [
+            ("BUZZ_MEDIA_UPLOAD_IP_HEADER", &self.upload_ip_header),
+            ("BUZZ_MEDIA_UPLOAD_PORT_HEADER", &self.upload_port_header),
+        ] {
+            if let Some(h) = value {
+                if h.is_empty() || !h.chars().all(|c| c.is_ascii_graphic() && c != ':') {
+                    return Err(format!("{name} is not a valid header name: {h:?}"));
```
Comment (wpfleger96) [inline crates/buzz-media/src/config.rs:116]: small one — this accepts any ASCII graphic except `:`, so a malformed name like `bad/header` or `bad,header` passes startup validation and then just fails empty at header lookup. That quietly weakens the otherwise-nice fail-loud config posture (the whole point of the coherence checks above is to catch operator mistakes at startup). Could validate with `http::HeaderName::from_bytes` instead — ideally storing the parsed `HeaderName` rather than a string so the lookup can't diverge from what you validated.

**Review by wpfleger96 — CHANGES_REQUESTED** [review]
Reviewed this with a second independent pass — flagging two blocking items and two smaller ones inline. The feature's well-built overall: off-by-default, fail-empty IP handling, fail-loud config coherence, tight privacy blast radius, and a good test matrix. My main concern is the fresh-upload ordering (comment on upload.rs) — a record-write failure can leave media servable via the sidecar with the moderation hook never firing, which is the exact case this feature exists to prevent. The clippy one is a hard CI fail. The other two are non-blocking but I'd like the IP-range tightening in before this ships enabled anywhere.

One thing we looked at and explicitly did NOT block on: the unbounded _uploads/ record fan-out per blob under repeated re-upload. Auth + the existing upload rate limiter run first, and the unit of moderation is the accepted event — so it's a cost/noise vector, not a correctness bug. Worth pairing with retention/lifecycle if cost ever bites.

**Review by kalvinnchau — APPROVED** [review]
